### PR TITLE
hostdb: Replace the last TSHashTable with IntrusiveHashMap.

### PR DESCRIPTION
**Pull request by chitianhao**
This is the last TSHashTable in use.

### CODE REVIEW

File: iocore/hostdb/HostDB.cc [hunk 1]
```diff
@@ -1558,14 +1558,13 @@ HostDBContinuation::iterateEvent(int event, Event *e)
       return EVENT_CONT;
     }
 
-    TSHashTable<RefCountCacheHashing> *partMap = hostDB.refcountcache->get_partition(current_iterate_pos).get_map();
-    for (RefCountCachePartition<HostDBInfo>::iterator_type i = partMap->begin(); i != partMap->end(); ++i) {
-      HostDBInfo *r = (HostDBInfo *)i.m_value->item.get();
+    IntrusiveHashMap<RefCountCacheLinkage> &partMap = hostDB.refcountcache->get_partition(current_iterate_pos).get_map();
+    for (auto &&it : partMap) {
```
Comment (SolidWallOfCode) [inline iocore/hostdb/HostDB.cc]: @ywkaras says "use `const&` not `&&`".

File: iocore/hostdb/HostDB.cc [hunk 2]
```diff
@@ -1558,14 +1558,13 @@ HostDBContinuation::iterateEvent(int event, Event *e)
...
+      HostDBInfo *r = (HostDBInfo *)it.item.get();
```
Comment (SolidWallOfCode) [inline iocore/hostdb/HostDB.cc]: No C-style casts.

File: iocore/hostdb/P_RefCountCache.h [hunk 3]
```diff
@@ -149,15 +158,12 @@ template <class C> class RefCountCachePartition
...
+  void dealloc_entry(IntrusiveHashMap<RefCountCacheLinkage>::iterator ptr);
 
   size_t count() const;
   void copy(std::vector<RefCountCacheHashEntry *> &items);
 
-  typedef typename TSHashTable<RefCountCacheHashing>::iterator iterator_type;
-  typedef typename TSHashTable<RefCountCacheHashing>::self hash_type;
-  typedef typename TSHashTable<RefCountCacheHashing>::Location location_type;
-  TSHashTable<RefCountCacheHashing> *get_map();
+  IntrusiveHashMap<RefCountCacheLinkage> &get_map();
```
Comment (SolidWallOfCode) [inline iocore/hostdb/P_RefCountCache.h]: I'd keep the `hash_type` here for clarity later - e.g.
```
using hash_type = IntrusiveHashMap<RefCountCacheLinkage>;
```

File: iocore/hostdb/P_RefCountCache.h [hunk 4]
```diff
@@ -188,11 +194,10 @@ Ptr<C>
 RefCountCachePartition<C>::get(uint64_t key)
 {
   this->metric_inc(refcountcache_total_lookups_stat, 1);
-  location_type l = this->item_map.find(key);
-  if (l.isValid()) {
+  if (auto it = this->item_map.find(key); it != this->item_map.end()) {
     // found
     this->metric_inc(refcountcache_total_hits_stat, 1);
-    return make_ptr((C *)l.m_value->item.get());
+    return make_ptr((C *)it->item.get());
```
Comment (SolidWallOfCode) [inline iocore/hostdb/P_RefCountCache.h]: No C-style casts.

File: iocore/hostdb/P_RefCountCache.h [hunk 5]
```diff
@@ -286,12 +284,11 @@ RefCountCachePartition<C>::clear()
...
   // hash elements and deallocate them, let alone remove them from the hash.
   // Hence, this monstrosity.
-  while (this->item_map.count() > 0) {
-    location_type pos = this->item_map.find(this->item_map.begin().m_value);
-
-    ink_assert(pos.isValid());
-    this->item_map.remove(pos);
-    this->dealloc_entry(pos);
+  for (auto it = this->item_map.begin(); it != this->item_map.end();) {
+    auto cur = it;
```
Comment (SolidWallOfCode) [inline iocore/hostdb/P_RefCountCache.h]: Here's a rare place where post-increment is the right choice.
```
auto cur = it++;
```

**Comment by chitianhao** [comment]
Updated. @SolidWallOfCode

File: iocore/hostdb/P_RefCountCache.h [hunk 6]
```diff
@@ -286,12 +286,10 @@ RefCountCachePartition<C>::clear()
...
   // hash elements and deallocate them, let alone remove them from the hash.
   // Hence, this monstrosity.
-  while (this->item_map.count() > 0) {
-    location_type pos = this->item_map.find(this->item_map.begin().m_value);
-
-    ink_assert(pos.isValid());
-    this->item_map.remove(pos);
-    this->dealloc_entry(pos);
+  for (auto it = this->item_map.begin(); it != this->item_map.end();) {
+    auto cur = it++;
```
Comment (zwoop) [inline iocore/hostdb/P_RefCountCache.h]: Why is this needed? Can’t you increment in the for loop?

File: iocore/hostdb/P_RefCountCache.h [hunk 7]
```diff
@@ -188,11 +196,10 @@ Ptr<C>
 RefCountCachePartition<C>::get(uint64_t key)
 {
   this->metric_inc(refcountcache_total_lookups_stat, 1);
-  location_type l = this->item_map.find(key);
-  if (l.isValid()) {
+  if (auto it = this->item_map.find(key); it != this->item_map.end()) {
```
Comment (zwoop) [inline iocore/hostdb/P_RefCountCache.h]: Why is this a for loop? It never iterates?

File: iocore/hostdb/P_RefCountCache.h [hunk 8]
```diff
@@ -239,44 +246,37 @@ template <class C>
 void
 RefCountCachePartition<C>::erase(uint64_t key, ink_time_t expiry_time)
 {
-  location_type l = this->item_map.find(key);
-  if (l.isValid()) {
-    if (expiry_time >= 0 && l.m_value->meta.expiry_time != expiry_time) {
+  if (auto it = this->item_map.find(key); it != this->item_map.end()) {
```
Comment (zwoop) [inline iocore/hostdb/P_RefCountCache.h]: Why is this a for loop? It seems without incrementing the iterator, this could in theory loop forever?

**Comment by zwoop** [comment]
Ignore the comments on for loops. I find that if syntax really confusing to read, since the internals looks like a for loops :-/

**Comment by zwoop** [comment]
Ok, I yield, just use 

```
if (auto it; ...)
```

when using the STL containers like this...

File: iocore/hostdb/P_RefCountCache.h (hunk 5 above)
Comment (chitianhao) [reply iocore/hostdb/P_RefCountCache.h]: Yeah, I think making it this way is better: 
```
  auto it = this->item_map.begin();
  while (it != this->item_map.end()) {
    auto cur = it;
    it       = this->item_map.erase(it);
    this->dealloc_entry(cur);
  }
```

**Comment by duke8253** [comment]
[approve ci autest]

File: iocore/hostdb/P_RefCountCache.h (hunk 5 above)
Comment (SolidWallOfCode) [reply iocore/hostdb/P_RefCountCache.h]: @zwoop - that doesn't help because `cur` and `it` still point at the same element, the destruction of which is the issue. The increment must happen _before_ the destruction, which means doing it after the `for` body is too late, regardless of how you copy the iterator and/or pointer around.

The better long term solution is to convert this to `IntrusiveDList` instead of being hand rolled and then use `IntrusiveDList::apply`, e.g. `item_map.apply([](Item* item) { delete item; });`.
